---
title: "Introduction to Analysis of high-throughput data using Bioconductor"
author: "Mark Dunning"
date: "February 2016"
output: html_document
---

## The Bioconductor project

![BioC](images/logo_bioconductor.png)

-  Packages analyse all kinds of Genomic data (>800)
- Compulsory documentation (*vignettes*) for each package
- 6-month release cycle
- Course Materials
- Example data and workflows
- Common, re-usable framework and functionality
- Available Support

## Example packages

![citations](images/citations.png)

##Downloading a package

Each package has its own landing page. e.g. http://bioconductor.org/packages/release/bioc/html/beadarray.html. Here you'll find;

- Installation script (will install all dependancies)
- Vignettes and manuals
- Details of package maintainer
- After downloading, you can load using the `library` function. e.g. `library(beadarray)`

##Reading data using Bioconductor

Recall that data can be read into R using `read.csv`, `read.delim`, `read.table` etc. Several packages provided special modifications of these to read raw data from different manufacturers

- `limma` for various two-colour platforms
- `affy` for Affymetrix data
- `beadarray`, `lumi`, `limma` for Illumina BeadArray data
- A common class is used to represent the data

##Reading data using Bioconductor

A dataset may be split into different components

- Matrix of expression values
- Sample information
- Annotation for the probes

In Bioconductor we will often put these data the same object for easy referencing. The `Biobase` package has all the code to do this.

## Example data

```{r results='hide',echo=FALSE}
suppressPackageStartupMessages(library(Biobase))
```

```{r}
library(Biobase)
data(sample.ExpressionSet)
sample.ExpressionSet
```

##Extracting data

Convenient `accessor` functions are provided
```{r}
evals <- exprs(sample.ExpressionSet)
dim(evals)
evals[1:4,1:3]
```

##Extracting data

Key to note the *rows* in the sample information are in the same order as the *columns* in the expression matrix

```{r}
sampleMat <- pData(sample.ExpressionSet)
dim(sampleMat)
head(sampleMat)
```



## Subsetting rules
`ExpressionSet` objects are designed to behave like data frames. e.g. to subset the first 10 genes


```{r}
sample.ExpressionSet[1:10,]
```

## Subsetting rules

What does this do?

```{r}
sample.ExpressionSet[,1:10]
```

## Subsetting rules

```{r}
males <- sampleMat[,1] == "Male"
sample.ExpressionSet[,males]
```

## Subsetting rules

```{r}
sample.ExpressionSet[,
        sampleMat$score < 0.5
        ]
```

## Starting to visualise the data

Recall that several plots can be created from a *vector* of numerical values

```{r}
hist(evals[,1])
```

## Starting to visualise the data

Or from a data frame

```{r}
boxplot(evals[,1:5])
```

## Starting to visualise the data

One sample against another 
```{r}
plot(evals[,1],evals[,2])
```

## Starting to visualise the data

One gene against another
```{r}
plot(evals[1,],evals[2,])
```


## The MA plot

We often work with **M** and **A** values as defined


```{r warning=FALSE}
M <- log2(evals[,1]) - log2(evals[,2])
A <- 0.5*(log2(evals[,1]) + log2(evals[,2]))
plot(A,M)
```

## The MA plot

- log transformation is used to put values on scale 0 to 16
- Line *M=0* indicates equivalent expression in two arrays
- Outliers on y axis are *candidates* to be *differentially expressed*
```{r echo=FALSE,warning=FALSE}
M <- log2(evals[,1]) - log2(evals[,2])
A <- 0.5*(log2(evals[,1]) + log2(evals[,2]))
plot(A,M)
abline(h=0,lty=2,col="red")
points(A[which(abs(M) > 3.5)], M[which(abs(M) > 3.5)],col="red",pch=20)
```

## The imageplot

Bioconductor packages will often provide a means to visualise the array surface as an `imageplot`

```{r echo=FALSE, eval=FALSE}


png("images/artefacts.png",width=800,height=400)
par(mfrow=c(1,2))
library(limma) 
datadir <- system.file("swirldata", package="marray") 
targets <- readTargets(file.path(datadir, "SwirlSample.txt")) 
RG <- read.maimages(file.path(datadir,targets$Names),source="spot") 
RG$genes <- readGAL(file.path(datadir, "fish.gal"))
 RG$printer <- getLayout(RG$genes)
MA <- normalizeWithinArrays(RG, method="none")
imageplot(MA$M[,1], RG$printer, zlim=c(-3,3))

load("rawdata.rda")

library(affyPLM)

plmset = fitPLM(rawdata)
image(plmset,60)
dev.off()
```

![badaffy](images/artefacts.png)



# R programming basics

##The working directory

```{r eval=FALSE}
getwd()
```


## Automating repetitive tasks

Often we want to repeat the same plot for a number of different arrays

- Histograms of various arrays 
```{r eval=FALSE}
hist(evals[,1])
hist(evals[,2])
hist(evals[,3])
....
```

- Note that each line of code is the same except for the column index

##Using a `for` loop

- We can simplify this code to 

```{r eval=FALSE}
hist(evals[,i])
```

- Where i can be 1, 2, or 3. 

##Using a `for` loop

```{r eval=FALSE}
i <- 1
hist(evals[,i])
i <- 2
hist(evals[,i])
i <- 3
hist(evals[,i])

```

##Using a `for` loop

- A loop can defined as follows. The code inside the `{}` will be run for each value of i in turn
```{r eval=FALSE}

for(i in 1:3){
  hist(evals[,i])
  }

```

##Using a `for` loop

- Multiple lines of code can be included in inside the `{}`
```{r eval=FALSE}

for(i in 1:3){
  ...
  hist(evals[,i])
  ...
  ...
  ...
  
  }

```



## Branching logic

- `if` and `else`

# Quality Assessment for microarrays

## Typical plots

- Boxplots of array distributions

```{r echo=FALSE,warning=FALSE,fig.height=4,fig.width=8}
load("rawdata.rda")
boxplot(rawdata[,1:10])
```

## Typical plots 

- MA plots

```{r echo=FALSE,fig.height=4,fig.width=8}
load("rawdata.rda")
suppressPackageStartupMessages(library(affyPLM))
par(mfrow=c(1,3))
MAplot(rawdata[,1:3])
```


##Typical plots

- Density plots

```{r echo=FALSE,warning=FALSE}
suppressPackageStartupMessages(library(affy))
plotDensity(log2(evals))
```


## Typical plots

- Clustering can reveal sample groupings
- (see tomorrow for more details)
```{r echo=FALSE}
d <- dist(t(evals))
plot(hclust(d),labels = sampleMat$type)
```


## Helpful packages

- affyPLM
- affyQCReport
- arrayQualityMetrics - example report >>[here](http://infosys.informatik.uni-mainz.de/research/timeseries-visualisation/array-quality-metrics)<<


# Normalisation

## Motivation

- We want to be observing *biological* and not *technical* variation
- We wouldn't expect such wholesale changes on a per-sample basis
- Easy option would to scale values for each array to median level
```{r echo=FALSE,warning=FALSE,fig.height=4,fig.width=8}
load("rawdata.rda")
boxplot(rawdata[,1:10])

x <- rawdata[,1:10]
         tmp <- unlist(indexProbes(x, which="both"))
         tmp <- tmp[seq(1, length(tmp), len = 5000)]
         df <- data.frame(log2(intensity(x)[tmp, ]))
        med <- median(as.matrix(df))

abline(h=med,lty=2,col="red")
```

## Simple scaling

- Genes on array 2 are on average `r median(df[,2]) - med` higher than the global median, so subtract `r median(df[,2]) - med` from each gene
- Genes on array 3 are on average `r abs(median(df[,3]) - med)` lower than the global median, so add `r abs(median(df[,3]) - med)` to each gene
- etc

```{r echo=FALSE,warning=FALSE,fig.height=4,fig.width=8}
load("rawdata.rda")
boxplot(rawdata[,1:10])
abline(h=med,lty=2,col="red")
```

## Non-linear effects

- We often compare to an *average* array which is the result of averaging each gene

```{r eval=FALSE}
aveArray <- rowMeans(data)
```

- Different effects can be seen when comparing to this theoretical array

```{r echo=FALSE,warning=FALSE,fig.height=4,fig.width=8,message=FALSE}
par(mfrow=c(1,3))
meanArray <- rowMeans(df)

M <- df[,1] - meanArray
A <- 0.5*(df[,1] + meanArray)
smoothScatter(A,M,ylim=c(-3,3),main="Array 1",ylab="Difference from average array")
abline(h=0,col="red")
lo <- loess(M~A)
lines(approx(lowess(A, M, f = 0.1)),col="blue")
legend("topright", col=c("blue","red"),lty=1,legend=c("line of best fit","M=0"))

M <- df[,2] - meanArray
A <- 0.5*(df[,2] + meanArray)
smoothScatter(A,M,ylim=c(-3,3),main="Array 2",ylab="Difference from average array")
abline(h=0,col="red")
lo <- loess(M~A)
lines(approx(lowess(A, M, f = 0.1)),col="blue")
legend("topright", col=c("blue","red"),lty=1,legend=c("line of best fit","M=0"))

M <- df[,3] - meanArray
A <- 0.5*(df[,3] + meanArray)
smoothScatter(A,M,ylim=c(-3,3),main="Array 3",ylab="Difference from average array")
abline(h=0,col="red")
lo <- loess(M~A)
lines(approx(lowess(A, M, f = 0.1)),col="blue")
legend("topright", col=c("blue","red"),lty=1,legend=c("line of best fit","M=0"))


```

## Quantile normalisation
Consider the following matrix of values to be normalised

```{r echo=FALSE}
df <- data.frame(Array1 = c(1,3,9,2,4), Array2 = c(3,4,2,1,9), Array3 = c(9,1,5,7,6))
rownames(df) <- LETTERS[1:nrow(df)]
df
```
Genes A, B, C, D and E measured on three arrays

## Quantile normalisation

Determine ranks of each column
```{r echo=FALSE}
df
```

```{r echo=FALSE}
rks <- apply(df, 2, function(x) paste("Rank",rank(x,ties.method="min"),sep=""))
rks
```

##Quantile normalisation

Sort each column Largest...Smallest

Original data
```{r echo=FALSE}
df
```
***
Sorted data
```{r echo=FALSE}
apply(df, 2,sort)
```
Then calculate target distribution by averaging the sorted rows
```{r echo=FALSE} 
target <- round(rowMeans(apply(df, 2,sort)),3)
names(target) <- paste("Rank", 1:length(target),sep="")
target
```


##Quantile normalisation

Go back to the rank matrix

```{r echo=FALSE}
rks

```
Substitue with values from the target distribution
```{r echo=FALSE}
target
```

```{r echo=FALSE}
rks[,1] <- gsub("Rank1",target["Rank1"],rks[,1])
rks
```

##Quantile normalisation

Go back to the rank matrix

```{r echo=FALSE}
rks

```
Substitue with values from the target distribution
```{r echo=FALSE}
target
```

```{r echo=FALSE}
rks[,1] <- gsub("Rank2",target["Rank2"],rks[,1])
rks
```


##Quantile normalisation

Go back to the rank matrix

```{r echo=FALSE}
rks

```
Substitue with values from the target distribution
```{r echo=FALSE}
target
```

```{r echo=FALSE}
rks[,1] <- gsub("Rank3",target["Rank3"],rks[,1])
rks
```

##Quantile normalisation

Go back to the rank matrix

```{r echo=FALSE}
rks

```
Substitue with values from the target distribution
```{r echo=FALSE}
target
```

```{r echo=FALSE}
rks[,1] <- gsub("Rank4",target["Rank4"],rks[,1])
rks
```

##Quantile normalisation

Go back to the rank matrix

```{r echo=FALSE}
rks

```
Substitue with values from the target distribution
```{r echo=FALSE}
target
```

```{r echo=FALSE}
rks[,1] <- gsub("Rank5",target["Rank5"],rks[,1])
rks
```

##Quantile normalisation
We then repeat to get the normalized matrix
```{r echo=FALSE}
for(i in 1:3){
rks[,i] <- gsub("Rank1",target["Rank1"],rks[,i])
rks[,i] <- gsub("Rank2",target["Rank2"],rks[,i])
rks[,i] <- gsub("Rank3",target["Rank3"],rks[,i])
rks[,i] <- gsub("Rank4",target["Rank4"],rks[,i])
rks[,i] <- gsub("Rank5",target["Rank5"],rks[,i])
}
rks <- as.data.frame(rks)
rownames(rks) <- rownames(df)
```

Original data
```{r echo=FALSE}
df
```
Normalised data

```{r echo=FALSE}
rks
```


##Final Code

```{r}
df <- data.frame(Array1 = c(1,3,9,2,4), 
                 Array2 = c(3,4,2,1,9), Array3 = c(9,1,5,7,6))
rownames(df) <- LETTERS[1:nrow(df)]
rks <- apply(df, 2, function(x) paste("Rank",
      rank(x,ties.method="min"),sep=""))
target <- round(rowMeans(apply(df, 2,sort)),3)
names(target) <- paste("Rank", 1:length(target),sep="")
for(i in 1:ncol(df)){
  for(nm in names(target)){
    rks[,i] <- gsub(nm,target[nm],rks[,i])  
      }
}
norm <- as.data.frame(rks)
```


##Effect of quantile normalisation

Caveats

- Distributions of samples are expected to be the same
- Majority of genes do not change between groups

```{r echo=FALSE,warning=FALSE,fig.height=4,fig.width=8,message=FALSE}

x <- rawdata[,1:10]
         tmp <- unlist(indexProbes(x, which="both"))
         tmp <- tmp[seq(1, length(tmp), len = 5000)]
         df <- data.frame(log2(intensity(x)[tmp, ]))
        med <- median(as.matrix(df))
par(mfrow=c(1,2))
boxplot(df,main="Before")
suppressPackageStartupMessages(library(limma))
boxplot(normalizeQuantiles(df),main="After")
```

## Next steps


- *Differential expression* to find genes that change between different experimental conditions
- *Clustering* to reveal groups in the data
- *Classification* to assign samples to pre-defined groups
- *Integrate* with another data type